internal/controller/ansibleRun/ansibleRun.go

```diff
@@ -527,3 +554,164 @@ func addBehaviorVars(pc *v1alpha1.ProviderConfig) map[string]string {
+			return err
+		}
+		return fmt.Errorf("failed to update lease: %w", err)
+	}
+
+	c.logger.Debug("updated lease", "lease", lease)
+	return nil
+}
+
+// Finds an available shard and acquires a lease for it. Will attempt to obtain one indefinitely.
+// This will also start a background go-routine to renew the lease continiously and release it when the process receives a shutdown signal
+func (c *connector) acquireAndHoldShard(o controller.Options, s SetupOptions) (int, error) {
+	ctx := context.Background()
+	acquiredLease := false
+	currentShard := -1
+
+	cfg := ctrl.GetConfigOrDie()
+	kube, err := client.New(cfg, client.Options{})
+	if err != nil {
+		return currentShard, err
+	}
+
+	for !acquiredLease {
+		for i := 0; i < s.ReplicasCount; i++ {
+			if err := c.acquireLease(ctx, kube, i); err == nil {
+				currentShard = i
+				o.Logger.Debug("acquired lease", "id", i)
+				acquiredLease = true
+				go func() {
+					sigHandler := ctrl.SetupSignalHandler()
+
+					for {
+						select {
+						case <-time.After(leaseRenewalInterval):
+							if err := c.acquireLease(ctx, kube, i); err != nil {
+								o.Logger.Info("failed to renew lease", "id", i, "err", err)
+								panic(err)
+							} else {
+								o.Logger.Debug("renewed lease", "id", i)
+							}
+						case <-sigHandler.Done():
+							o.Logger.Info("controller is shutting down, releasing lease")
+							if err := c.releaseLease(ctx, kube, i); err != nil {
+								o.Logger.Info("failed to release lease", "lease", err)
+							}
+							o.Logger.Debug("released lease")
+							s.ProgramCancel()
+							return
+						}
+					}
+				}()
+				// Lease is acquired and background goroutine started for renewal, we can safely break to return the current shard
+				break
+			} else {
+				o.Logger.Debug("cannot acquire lease", "id", i, "err", err)
+				time.Sleep(leaseAcquireAttemptInterval)
+			}
+		}
+	}
+
+	return currentShard, nil
+}
+
+// Define a predicate function to filter resources based on consistent hashing
+func isResourceForShard(targetShard, totalShards int) predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e crevent.CreateEvent) bool {
+			return isResourceForShardHelper(e.Object, targetShard, totalShards)
+		},
+		UpdateFunc: func(e crevent.UpdateEvent) bool {
+			return isResourceForShardHelper(e.ObjectNew, targetShard, totalShards)
+		},
+		DeleteFunc: func(e crevent.DeleteEvent) bool {
+			return isResourceForShardHelper(e.Object, targetShard, totalShards)
+		},
+		GenericFunc: func(e crevent.GenericEvent) bool {
+			return isResourceForShardHelper(e.Object, targetShard, totalShards)
+		},
+	}
+}
+
+// Helper function to check if the resource belongs to the current shard
+func isResourceForShardHelper(obj client.Object, targetShard, totalShards int) bool {
+	// Calculate a hash of the resource name
+	hash := hashString(obj.GetName())
+	// Perform modulo operation to determine the shard
+	shard := hash % uint32(totalShards)
```

Comment:
this is not `consistent hash` but `modular hashing` (c.f https://www.geeksforgeeks.org/regular-hashing-vs-consistent-hashing/#what-is-regular-hashing) . it is ok to start with `modular hashing` but in the long run I prefer `consistent hash` to avoid significant data remapping when replicas num changes.

Reply:
if you are not planning to switch to `consistent hash` in this PR please add a comment regarding this future refacto